File: src/utils/prime.js
export function isPrime(n) {
  if (typeof n !== 'number' || !Number.isInteger(n)) return false;
  if (n <= 1) return false;
  if (n <= 3) return true;
  if (n % 2 === 0 || n % 3 === 0) return false;

  for (let i = 5; i * i <= n; i += 6) {
    if (n % i === 0 || n % (i + 2) === 0) {
      return false;
    }
  }

  return true;
}

export function primesUpTo(limit) {
  if (typeof limit !== 'number' || !Number.isInteger(limit) || limit < 2) return [];

  const sieve = new Array(limit + 1).fill(true);
  sieve[0] = false;
  sieve[1] = false;

  for (let p = 2; p * p <= limit; p++) {
    if (sieve[p]) {
      for (let multiple = p * p; multiple <= limit; multiple += p) {
        sieve[multiple] = false;
      }
    }
  }

  return sieve
    .map((isPossiblePrime, index) => (isPossiblePrime ? index : null))
    .filter((value) => value !== null);
}

export function nextPrimeAfter(n) {
  let candidate = Math.max(2, Math.floor(n) + 1);
  while (!isPrime(candidate)) {
    candidate += 1;
  }
  return candidate;
}

File: prime.py
def is_prime(n: int) -> bool:
    """Return True if n is a prime number."""
    if not isinstance(n, int) or n <= 1:
        return False
    if n <= 3:
        return True
    if n % 2 == 0 or n % 3 == 0:
        return False

    i = 5
    while i * i <= n:
        if n % i == 0 or n % (i + 2) == 0:
            return False
        i += 6
    return True


def primes_up_to(limit: int) -> list[int]:
    """Return a list of prime numbers from 2 up to limit."""
    if not isinstance(limit, int) or limit < 2:
        return []

    sieve = [True] * (limit + 1)
    sieve[0:2] = [False, False]

    for number in range(2, int(limit**0.5) + 1):
        if sieve[number]:
            for multiple in range(number * number, limit + 1, number):
                sieve[multiple] = False

    return [i for i, prime in enumerate(sieve) if prime]


if __name__ == '__main__':
    n = 31
    print(f"is_prime({n}) = {is_prime(n)}")
    print(f"primes_up_to(50) = {primes_up_to(50)}")
